src/system/rpi_pico/system.c

```diff
@@ -0,0 +1,312 @@
+    z_arg->arg = arg;
+    z_arg->join_event = task->join_event = xEventGroupCreate();
+
+    if (attr == NULL) {
+        attr = &z_default_task_attr;
+    }
+
+    if (xTaskCreate(z_task_wrapper, attr->name, attr->stack_depth, z_arg, attr->priority, &task->handle) != pdPASS) {
+        return -1;
+    }
+
+    return 0;
+}
+
+z_result_t _z_task_join(_z_task_t *task) {
+    xEventGroupWaitBits(task->join_event, 1, pdFALSE, pdFALSE, portMAX_DELAY);
+    return 0;
+}
+
+z_result_t _z_task_detach(_z_task_t *task) {
+    _ZP_UNUSED(task);
+    assert(false);
+    return _Z_ERR_GENERIC;
+}
+
+z_result_t _z_task_cancel(_z_task_t *task) {
+    vTaskDelete(task->handle);
+    return 0;
+}
+
+void _z_task_free(_z_task_t **task) {
+    z_free((*task)->join_event);
+    z_free(*task);
+}
+
+/*------------------ Mutex ------------------*/
+z_result_t _z_mutex_init(_z_mutex_t *m) {
+    *m = xSemaphoreCreateRecursiveMutex();
+    return *m == NULL ? -1 : 0;
+}
+
+z_result_t _z_mutex_drop(_z_mutex_t *m) {
+    z_free(*m);
+    return 0;
+}
+
+z_result_t _z_mutex_lock(_z_mutex_t *m) { return xSemaphoreTakeRecursive(*m, portMAX_DELAY) == pdTRUE ? 0 : -1; }
+
+z_result_t _z_mutex_try_lock(_z_mutex_t *m) { return xSemaphoreTakeRecursive(*m, 0) == pdTRUE ? 0 : -1; }
+
+z_result_t _z_mutex_unlock(_z_mutex_t *m) { return xSemaphoreGiveRecursive(*m) == pdTRUE ? 0 : -1; }
+
+/*------------------ CondVar ------------------*/
+static UBaseType_t CONDVAR_MAX_WAITERS_COUNT = 255;
+
+z_result_t _z_condvar_init(_z_condvar_t *cv) {
+    if (!cv) {
+        return _Z_ERR_GENERIC;
+    }
+    cv->mutex = xSemaphoreCreateMutex();
+    cv->sem = xSemaphoreCreateCounting(CONDVAR_MAX_WAITERS_COUNT, 0);
+    cv->waiters = 0;
+
+    if (!cv->mutex || !cv->sem) {
+        return _Z_ERR_GENERIC;
+    }
+    return _Z_RES_OK;
+}
+
+z_result_t _z_condvar_drop(_z_condvar_t *cv) {
+    if (!cv) {
+        return _Z_ERR_GENERIC;
+    }
+    vSemaphoreDelete(cv->sem);
+    vSemaphoreDelete(cv->mutex);
+    return _Z_RES_OK;
+}
+
+z_result_t _z_condvar_signal(_z_condvar_t *cv) {
+    if (!cv) {
+        return _Z_ERR_GENERIC;
+    }
+
+    xSemaphoreTake(cv->mutex, portMAX_DELAY);
+    if (cv->waiters > 0) {
+        xSemaphoreGive(cv->sem);
+        cv->waiters--;
+    }
+    xSemaphoreGive(cv->mutex);
+
+    return _Z_RES_OK;
+}
+
+z_result_t _z_condvar_signal_all(_z_condvar_t *cv) {
+    if (!cv) {
+        return _Z_ERR_GENERIC;
+    }
+
+    xSemaphoreTake(cv->mutex, portMAX_DELAY);
+    while (cv->waiters > 0) {
+        xSemaphoreGive(cv->sem);
+        cv->waiters--;
+    }
+    xSemaphoreGive(cv->mutex);
+
+    return _Z_RES_OK;
+}
+
+z_result_t _z_condvar_wait(_z_condvar_t *cv, _z_mutex_t *m) {
+    if (!cv || !m) {
+        return _Z_ERR_GENERIC;
+    }
+
+    xSemaphoreTake(cv->mutex, portMAX_DELAY);
+    cv->waiters++;
+    xSemaphoreGive(cv->mutex);
+
+    xSemaphoreGive(*m);
+
+    xSemaphoreTake(cv->sem, portMAX_DELAY);
+
+    xSemaphoreTake(*m, portMAX_DELAY);
```

Comment:
`_z_mutex_t` is a recursive mutex so we can't use `xSemaphoreGive`. Why not just use `_z_mutex_lock/unlock`?
```suggestion
    _z_mutex_unlock(m);

    xSemaphoreTake(cv->sem, portMAX_DELAY);

    _z_mutex_lock(m);
```

Reply:
Good catch! Thanks!